### PR TITLE
Crypto.com Margin

### LINKED ISSUE #14042: Request: margin trade for cryptocom (crypto.com)
As title, could you implement margin trade for crypto.com?
There are APIs for margin trading, seeing API document.

### PR DESCRIPTION
**Pull request by Dan-krm**
Adds margin functionality to Crypto.com:
fixes #14042

### Methods to be completed:
- [x] transfer
- [x] fetchTransfers
- [x] fetchBalance
- [x] createOrder
- [x] fetchOrders
- [x] fetchOpenOrders
- [x] fetchOrder
- [x] cancelOrder
- [x] cancelAllOrders
- [x] fetchMyTrades
- [x] borrowMargin
- [x] repayMargin
- [x] fetchBorrowInterest
- [x] fetchBorrowRates
- [ ] ~fetchBorrowRate~
- [ ] ~fetchBorrowHistory~
- [ ] ~fetchBorrowHistories~
- [ ] ~setMarginMode~

### transfer:
```
cryptocom.transfer (USDT, 16, spot, margin)
2022-06-29T06:02:03.453Z iteration 0 passed in 279 ms

{
  info: { id: '1656482523210', method: 'private/margin/transfer', code: '0' },
  id: '1656482523210',
  timestamp: undefined,
  datetime: undefined,
  currency: undefined,
  amount: undefined,
  fromAccount: undefined,
  toAccount: undefined,
  status: undefined
}
2022-06-29T06:02:03.453Z iteration 1 passed in 279 ms
```

### fetchTransfers:
```
node examples/js/cli cryptocom fetchTransfers USDT undefined undefined '{"direction":"OUT"}'

cryptocom.fetchTransfers (USDT, , , [object Object])
2022-06-29T06:46:47.512Z iteration 0 passed in 278 ms

           id |     timestamp |                 datetime | currency | amount | fromAccount | toAccount | status | direction
---------------------------------------------------------------------------------------------------------------------------
1656485207268 | 1656482503867 | 2022-06-29T06:01:43.867Z |     USDT |     16 |        spot |    margin |     ok |       OUT
1 objects
2022-06-29T06:46:47.512Z iteration 1 passed in 278 ms
```

### fetchBalance:
```
cryptocom.fetchBalance ()
2022-06-29T19:39:21.911Z iteration 0 passed in 286 ms

{
    {
        {
          balance: '0',
          available: '0',
          order: '0',
          borrowed: '0',
          position: '0',
          positionHomeCurrency: '0',
          positionBtc: '0',
          lastPriceHomeCurrency: '0.05115',
          lastPriceBtc: '0.00000254',
          currency: 'SKL',
          accrued_interest: '0',
          liquidation_price: '0'
        },
        ...
      ],
      is_liquidating: false,
      total_balance: '16',
      total_balance_btc: '0.00079556',
      equity_value: '16',
      equity_value_btc: '0.00079556',
      total_borrowed: '0',
      total_borrowed_btc: '0',
      total_accrued_interest: '0',
      total_accrued_interest_btc: '0',
      margin_score: 'GOOD',
      currency: 'USDT'
    }
  },
  BTC: { free: 0, used: 0, total: 0 },
  USDT: { free: 16, used: 0, total: 16 },
  CRO: { free: 0, used: 0, total: 0 },
 ...
  free: {
    BTC: 0,
    USDT: 16,
    CRO: 0,
    ...
  },
  used: {
    BTC: 0,
    USDT: 0,
    CRO: 0,
    ...
  },
  total: {
    BTC: 0,
    USDT: 16,
    CRO: 0,
    ...
  }
}
```

### createOrder:
```
cryptocom.createOrder (BTC/USDT, limit, buy, 0.0005, 18000)
2022-06-29T19:39:21.911Z iteration 0 passed in 286 ms

{
  info: { order_id: '2640558065979074241' },
  id: '2640558065979074241',
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: 'BTC/USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  cost: undefined,
  fee: undefined,
  average: undefined,
  trades: [],
  fees: []
}
2022-06-29T19:39:21.911Z iteration 1 passed in 286 ms
```

### fetchOrders:
```
cryptocom.fetchOrders (BTC/USDT)
2022-06-29T19:49:06.807Z iteration 0 passed in 286 ms

                 id | clientOrderId |     timestamp |                 datetime | lastTradeTimestamp |   status |   symbol |  type | timeInForce | postOnly | side | price | amount | filled | remaining | cost | fee | average | trades | fees
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2640558065979074241 |               | 1656531562766 | 2022-06-29T19:39:22.766Z |      1656531562766 | rejected | BTC/USDT | limit |         GTC |          |  buy | 18000 | 0.0005 |      0 |    0.0005 |    0 |     |         |     [] |   []
2640571243535947235 |               | 1656531955488 | 2022-06-29T19:45:55.488Z |      1656531955488 | rejected | BTC/USDT | limit |         GTC |          |  buy | 18000 | 0.0005 |      0 |    0.0005 |    0 |     |         |     [] |   []
2 objects
2022-06-29T19:49:06.807Z iteration 1 passed in 286 ms
```

### fetchOpenOrders:
```
cryptocom.fetchOpenOrders (BTC/USDT)
2022-06-29T20:03:23.498Z iteration 0 passed in 267 ms

                 id | clientOrderId |     timestamp |                 datetime | lastTradeTimestamp | status |   symbol |  type | timeInForce | postOnly | side | price | amount | filled | remaining | cost | fee | average | trades | fees
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2640600951149394145 |               | 1656532840843 | 2022-06-29T20:00:40.843Z |      1656532840843 |   open | BTC/USDT | limit |         GTC |          |  buy | 15000 | 0.0001 |      0 |    0.0001 |    0 |     |         |     [] |   []
1 objects
2022-06-29T20:03:23.498Z iteration 1 passed in 267 ms
```

### fetchOrder:
```
node examples/js/cli cryptocom fetchOrder "'2640600951149394145'" --verbose

cryptocom.fetchOrder (2640600951149394145)
fetch Request:
 cryptocom POST https://api.crypto.com/v2/private/margin/get-order-detail

ResponseBody:
 {"id":1656533261392,"method":"private/margin/get-order-detail","code":0,"result":{}}

2022-06-29T20:07:41.736Z iteration 0 passed in 379 ms

{
  info: {},
  id: undefined,
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: undefined,
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  cost: undefined,
  fee: undefined,
  average: undefined,
  trades: [],
  fees: []
}
2022-06-29T20:07:41.736Z iteration 1 passed in 379 ms
```

### cancelOrder:
```
cryptocom.cancelOrder (2640600951149394145, BTC/USDT)
2022-06-29T20:13:09.040Z iteration 0 passed in 287 ms

{
  info: { id: '1656533588785', method: 'private/cancel-order', code: '0' },
  id: undefined,
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  status: undefined,
  symbol: undefined,
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  cost: undefined,
  fee: undefined,
  average: undefined,
  trades: [],
  fees: []
}
2022-06-29T20:13:09.040Z iteration 1 passed in 287 ms
```

### cancelAllOrders:
```
cryptocom.cancelAllOrders (BTC/USDT)
2022-06-29T20:16:21.353Z iteration 0 passed in 290 ms

{ id: '1656533781098', method: 'private/cancel-all-orders', code: '0' }
2022-06-29T20:16:21.353Z iteration 1 passed in 290 ms
```

### fetchMyTrades:
```
cryptocom.fetchMyTrades (BTC/USDT)
2022-06-29T20:35:57.195Z iteration 0 passed in 293 ms

                 id |     timestamp |                 datetime |   symbol | side |    price |  amount |      cost |               order | takerOrMaker | type |                            fee |                             fees
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2640669883735463296 | 1656534895194 | 2022-06-29T20:34:55.194Z | BTC/USDT |  buy | 20200.26 | 0.00001 | 0.2020026 | 2640669883630473091 |        taker |      | {"currency":"BTC","cost":4e-8} | [{"currency":"BTC","cost":4e-8}]
1 objects
2022-06-29T20:35:57.195Z iteration 1 passed in 293 ms
```

### borrowMargin:
```
cryptocom.borrowMargin (USDT, 30)
2022-06-30T20:06:18.892Z iteration 0 passed in 367 ms

{
  id: 1656619578559,
  currency: 'USDT',
  amount: 30,
  symbol: undefined,
  timestamp: undefined,
  datetime: undefined,
  info: { id: '1656619578559', method: 'private/margin/borrow', code: '0' }
}
2022-06-30T20:06:18.892Z iteration 1 passed in 367 ms
```

### repayMargin:
```
cryptocom.repayMargin (USDT, 30)
2022-06-30T20:12:57.803Z iteration 0 passed in 318 ms

{
  id: 1656619977521,
  currency: 'USDT',
  amount: 30,
  symbol: undefined,
  timestamp: undefined,
  datetime: undefined,
  info: {
    id: '1656619977521',
    method: 'private/margin/repay',
    code: '0',
    result: { badDebt: '0' }
  }
}
2022-06-30T20:12:57.803Z iteration 1 passed in 318 ms
```

### fetchBorrowInterest:
```
cryptocom.fetchBorrowInterest ()
2022-07-01T20:31:37.451Z iteration 0 passed in 287 ms

symbol | marginMode | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------------------
       |            |     USDT |     4e-8 |     0.000025 |                | 1656706499559 | 2022-07-01T20:14:59.559Z
       |            |     USDT |     4e-8 |     0.000025 |                | 1656702899559 | 2022-07-01T19:14:59.559Z
       |            |     USDT |     4e-8 |     0.000025 |                | 1656699299559 | 2022-07-01T18:14:59.559Z
...
20 objects
2022-07-01T20:31:37.451Z iteration 1 passed in 287 ms
```

### fetchBorrowRates:
```
cryptocom.fetchBorrowRates ()
2022-07-01T20:52:16.864Z iteration 0 passed in 293 ms

currency |       rate |  period |     timestamp |                 datetime
--------------------------------------------------------------------------
    AGLD | 0.00003334 | 3600000 | 1656708736864 | 2022-07-01T20:52:16.864Z
   MATIC | 0.00003334 | 3600000 | 1656708736864 | 2022-07-01T20:52:16.864Z
    IOTX | 0.00003334 | 3600000 | 1656708736864 | 2022-07-01T20:52:16.864Z
...
130 objects
2022-07-01T20:52:16.864Z iteration 1 passed in 293 ms
```

### CODE REVIEW

File: js/cryptocom.js [hunk 1]
```diff
@@ -921,13 +968,14 @@ module.exports = class cryptocom extends Exchange {
         }
         const request = {};
         const [ marketType, query ] = this.handleMarketTypeAndParams ('fetchOrder', market, params);
-        if (marketType === 'spot') {
+        if ((marketType === 'spot') || (marketType === 'spot')) {
```
Comment (kjnh10) [inline js/cryptocom.js:971]: Is this a bug?
It seems right hand side should be marketType == 'margin'.

Comment (kjnh10) [reply js/cryptocom.js:971]: @kroitor

**Comment by kjnh10** [comment]
Hi, I have a question about how to use margin.
Like below code, I want margin mode for all requests once I set it, but not setting it every when I request.

```python
import pprint
import ccxt

exchange = ccxt.cryptocom()
exchange.options['defaultType'] = 'margin'  # I want margin mode for all request once I set.

currency = 'BTC/USDT'
order_result = exchange.create_order(
    symbol=currency,
    type='limit',
    side='sell',
    amount=1,
    price=50000,
    # params={'defaultType': 'margin'},  # I don't want to pass params every when request
    )

fetch_result = exchange.fetch_order(
    id=order_result['id'],
    symbol=currency
    # params={'defaultType': 'margin'},  # I don't want to pass params every when request
    )

```

But this code fails because requests are sent to spot API.

Is there a way to fit my demand?

**Comment by kjnh10** [comment]
image

In this function, the value set by exchange.options is overrides by market(currency)['type'].
Is this expected behavior?
I think it's natural the value set by exchange.options is prioritized.

File: js/cryptocom.js (hunk 1 above)
Comment (Dan-krm) [reply js/cryptocom.js:971]: #14107

**Comment by kjnh10** [comment]
@kroitor 
Could you check my comment above?

**Comment by kjnh10** [comment]
@kroitor 
Could you check my comment above?

**Pull request merged**